[DB-606] gRPC filtered $all subscriptions with smooth transitions between live and catchup

shaan1337 opened the pull request:
Added: gRPC filtered $all subscriptions with smooth transitions between live and catchup. Subscriptions no longer drop with "consumer too slow" reason.

YoEight commented on src/EventStore.Core/Services/Transport/Enumerators/Enumerator.AllSubscriptionFiltered.cs:
```diff
@@ -71,462 +59,380 @@ public class AllSubscriptionFiltered : IAsyncEnumerator<ReadResponse> {
+								confirmationPositionTcs.TrySetResult(new TFPos(caughtUp, caughtUp));
+								return;
+							case ClientMessage.SubscriptionDropped dropped:
+								Log.Debug(
```
I believe we should switch it to `Verbose` here.

shaan1337 replied:
same as [this comment](https://github.com/EventStore/EventStore/pull/4135#discussion_r1473614863), i'm open to the change, but i suggest we do it in another PR as it also affects the other subscription types

timothycoleman commented on src/EventStore.Core/Services/Transport/Enumerators/Enumerator.AllSubscriptionFiltered.cs:
```diff
@@ -71,462 +59,380 @@ public class AllSubscriptionFiltered : IAsyncEnumerator<ReadResponse> {
-						Log.Verbose(
-							"Subscription {subscriptionId} to $all:{eventFilter} skipping event {position}.",
-							_subscriptionId, _eventFilter, position);
+						// this should no longer happen
```
if it does happen then we should skip that event like we are doing

shaan1337 replied:
yes, previously it could occur by design but now it's just a fail-safe mechanism to make sure we don't send events out of order

timothycoleman commented on src/EventStore.Core/Services/Transport/Enumerators/Enumerator.AllSubscriptionFiltered.cs:
```diff
@@ -71,462 +59,380 @@ public class AllSubscriptionFiltered : IAsyncEnumerator<ReadResponse> {
+
+						if (message is not ClientMessage.FilteredReadAllEventsForwardCompleted completed)
+							throw ReadResponseException.UnknownMessage
+								.Create<ClientMessage.ReadAllEventsForwardCompleted>(message);
```
```suggestion
								.Create<ClientMessage.FilteredReadAllEventsForwardCompleted>(message);
```
probably just an oversight

shaan1337 replied:
thanks, yup i missed that

timothycoleman commented on src/EventStore.Core/Services/Transport/Enumerators/Enumerator.AllSubscriptionFiltered.cs:
```diff
@@ -71,462 +59,380 @@ public class AllSubscriptionFiltered : IAsyncEnumerator<ReadResponse> {
+					} else if (liveEvent.Checkpoint.HasValue) {
+						// this is a valid checkpoint that we can send towards the client. send it and update `checkpoint`
+						await SendCheckpointToSubscription(liveEvent.Checkpoint.Value, ct);
+						checkpoint = new TFPos(liveEvent.Checkpoint.Value.CommitPosition, liveEvent.Checkpoint.Value.PreparePosition);
```
```suggestion
						checkpoint = liveEvent.Checkpoint.Value;
```
any reason not to do this?

shaan1337 replied:
thanks, not sure why i wrote it that way :smile:

shaan1337 commented on src/EventStore.Core.Tests/Services/Transport/Enumerators/Enumerator.AllSubscriptionFiltered.CombinationTests.cs:
```diff
@@ -576,6 +564,25 @@ public enum EventFilterType {
+				CheckpointType.OneBeforeLast;
+
+			// or we were live and we've received `checkpoint interval` events since the last checkpoint
+			shouldCheckpoint |= !catchingUp && numEventsSinceLastCheckpoint == CheckpointInterval;
```
@YoEight hey, following the new semantics, the checkpoint interval is an upper bound, so `numEventsSinceLastCheckpoint must be <= CheckpointInterval`.

Say `CheckpointInterval` is 16, what the above check verifies is that if we've received 16 events since the last checkpoint, then the next thing we enumerate must be a checkpoint as we've already reached the upper bound limit.

shaan1337 commented on src/EventStore.Core.Tests/Services/Transport/Grpc/StreamsTests/SubscribeToAllFilteredTests.cs:
```diff
@@ -451,7 +451,15 @@ public when_subscribing_to_all_with_a_filter_and_transitioning_to_live(int catch
+				// this is not a strict requirement - the important thing being verified here is that the checkpoints are
+				// received after the correct interval.
+
+				_checkpointPositions.ForEach(p => Assert.AreEqual(_numberOfEventsToCatchUp % CheckpointInterval, p % CheckpointInterval, $"checkpoint at: {p}"));
```
hmmm yup you're right, with this assert we're not really verifying if the checkpoints are at the correct interval. i'll try to adjust the assertion

shaan1337 commented on src/EventStore.Core/Services/Transport/Enumerators/Enumerator.AllSubscriptionFiltered.cs:
```diff
@@ -71,462 +59,380 @@ public class AllSubscriptionFiltered : IAsyncEnumerator<ReadResponse> {
+							_subscriptionId, _eventFilter);
+					_channel.Writer.TryComplete(ex);
+				} finally {
+					Log.Information("Subscription {subscriptionId} to $all:{eventFilter} has ended.",
```
we've set it to `Information` for the two other subscription types too (`stream` & `$all`) so i suggest that do it  in another PR if we're going for the change.

right now we log at the `info` log level only when the subscription starts and ends. given that a subscription stays on permanently until the client dies, this should be relatively lightweight on the logs

shaan1337 commented on src/EventStore.Core/Services/Transport/Enumerators/Enumerator.AllSubscriptionFiltered.cs:
```diff
@@ -71,462 +59,380 @@ public class AllSubscriptionFiltered : IAsyncEnumerator<ReadResponse> {
+								foreach (var @event in completed.Events) {
+									var eventPosition = @event.OriginalPosition!.Value;
+
+									// this should be true only for the first event of the first page
```
yup :thinking: we could also do an equality check here to make the check more strict (as otherwise we will miss the "fail-safe" warning)

shaan1337 replied:
i've adjusted the comment for $all/filtered $all, getting rid of the `should` to avoid any doubts

shaan1337 commented:
@YoEight @timothycoleman i've addressed the review comments and also made the following change as it was possible to send the same checkpoint twice:

```
@@ -33,6 +33,7 @@ namespace EventStore.Core.Services.Transport.Enumerators {
                        private ReadResponse _current;
                        private bool _disposed;
                        private Position? _currentPosition;
+                       private Position? _currentCheckpoint;
 
                        public ReadResponse Current => _current;
                        public string SubscriptionId { get; }
@@ -112,9 +113,20 @@ namespace EventStore.Core.Services.Transport.Enumerators {
                                        _currentPosition = position;
                                } else if (readResponse is ReadResponse.CheckpointReceived checkpointReceived) {
                                        var checkpointPos = new Position(checkpointReceived.CommitPosition, checkpointReceived.PreparePosition);
+
+                                       if (_currentCheckpoint.HasValue && checkpointPos <= _currentCheckpoint.Value) {
+                                               // in some cases, it's possible to receive the same checkpoint twice for example:
+                                               // i) when the subscription goes live and it turns out that the next thing on the live channel is a checkpoint
+                                               // ii) when catching-up, reaching the checkpoint interval and the next page read reaches the end of the stream
+
+                                               goto ReadLoop;
+                                       }
+
                                        Log.Verbose(
                                                "Subscription {subscriptionId} to $all:{eventFilter} received checkpoint {position}.",
                                                _subscriptionId, _eventFilter, checkpointPos);
+
+                                       _currentCheckpoint = checkpointPos;
                                }
 
                                _current = readResponse;
```
